common/minimodel/__init__.py

```diff
@@ -1008,3 +1013,52 @@ def total_used(self) -> timedelta:
 
     def observations(self) -> List[Observation]:
         return self.root_group.observations()
+
+
+class Plans:
+    """
+    A collection of Nightly Plan from a corresponding Site
+    """
+    def __init__(self, night_events):
+        self.durations = [len(night_events.times[idx]) for idx, _ in enumerate(night_events.time_grid)]
+        self.start_times = [night_events.local_times[idx][0] for idx, _ in enumerate(night_events.time_grid)]
+        self.end_times = [night_events.local_times[idx][-1] for idx, _ in enumerate(night_events.time_grid)]
+        self.time_slot_length = night_events.time_slot_length
+
+    def __iter__(self):
+        self.plans = []
+        return self
+
+    def __next__(self):
+        if len(self.plans) <= len(self.durations):
+            self.plans.append(Plan(next(iter(self.start_times)),
+                                   next(iter(self.end_times)),
+                                   slot_length=self.time_slot_length,
+                                   night_duration=next(iter(self.durations))))
```

Comment:
I'm not clear why we are appending in `__next__`.
Why don't we set this up beforehand?
I don't think we should be mutating in an iterator method. It seems like a dangerous idea.

Reply:
If we iterate twice over, you'll be repeating work, by clearing out `plans` and then recreating it, no?

And if we have two iterators, it'll play havoc and the value will depend on who calls `next` first / how often.